vello_shaders/src/compile/msl.rs

```diff
@@ -0,0 +1,49 @@
+use naga::back::msl;
+
+use super::{BindType, ShaderInfo};
+
+pub fn translate(shader: &ShaderInfo) -> Result<String, msl::Error> {
+    let mut map = msl::PerStageMap::default();
+    let mut buffer_index = 0u8;
+    let mut image_index = 0u8;
+    let mut binding_map = msl::BindingMap::default();
+    for resource in &shader.bindings {
+        let binding = naga::ResourceBinding {
+            group: resource.location.0,
+            binding: resource.location.1,
+        };
+        let mut target = msl::BindTarget::default();
+        match resource.ty {
+            BindType::Buffer | BindType::BufReadOnly | BindType::Uniform => {
+                target.buffer = Some(buffer_index);
+                buffer_index += 1;
+            }
+            BindType::Image | BindType::ImageRead => {
+                target.texture = Some(image_index);
+                image_index += 1;
+            }
+        }
+        target.mutable = resource.ty.is_mutable();
+        binding_map.insert(binding, target);
+    }
+    map.cs = msl::PerStageResources {
```

Comment:
https://github.com/gfx-rs/naga/pull/2237 has been merged into mainline naga, which replaces `PerStageResources` with `EntryPointResources` and indexes them by entry point name. You can use that here if you pin naga against HEAD.

You should probably cache the entry-point name that the `ShaderInfo` is initialized with and use that here as the key for `EntryPointResources`.

Alternately you could process all reachable bindings for every entry-point in the module, set the map once and translate it to MSL in one go with the same `EntryPointResourceMap`. That might also make sense if we wanted a single `ShaderInfo` to support multiple entry-points.

Reply:
Thanks for the update. I've been waiting on that PR to continue iterating on this. 

I'd like to experiment with merging some of our shaders but am deferring that until multisampled path rendering lands to avoid creating unnecessary conflicts.

In the meantime, I'll go ahead with the changes you suggest here and also add a path to generate a C++ header as we discussed.